content/tutorials/numpy_integration/grass_numpy_integration.qmd

```diff
@@ -0,0 +1,253 @@
+
+This tutorial is prepared to be run in a Jupyter notebook locally
+or using services such as Google Colab. You can [download the notebook here](grass_numpy_integration.ipynb).
+
+If you are not sure how to get started with GRASS, checkout the tutorials [Get started with GRASS & Python in Jupyter Notebooks](../get_started/fast_track_grass_and_python.qmd)
+and [Get started with GRASS in Google Colab](../get_started/grass_gis_in_google_colab.qmd).
+
+:::
+
+![](logos.webp)
+
+# Generating a fractal surface in GRASS
+
+First we will import GRASS packages:
+
+```{python}
+# import standard Python packages
+import os
+import sys
+import subprocess
+
+sys.path.append(
+    subprocess.check_output(["grass", "--config", "python_path"], text=True).strip()
+)
+
+# import GRASS python packages
+import grass.script as gs
+import grass.jupyter as gj
+from grass.tools import Tools
+```
+
+Create a new GRASS project called "landlab". Since we will be working with artifical data, we don't need to
+provide any coordinate reference system, resulting in a generic cartesian system.
+
+```{python}
+gs.create_project("landlab")
+```
+
+Initialize a session in this project and create a `Tools` object we will use for calling GRASS tools:
+
+```{python}
+session = gj.init("landlab")
+tools = Tools()
+```
+
+Since we are generating artificial data, we need to specify the dimensions (number of rows and columns).
+We also need to let GRASS know the actual coordinates; we will do that by setting
+the [computational region](https://grass.osgeo.org/grass-stable/manuals/g.region.html). Lower-left (south-west) corner of the data will be at the coordinates (0, 0),
+the coordinates of the upper-right (nort-east) corner are number of rows times cell resolution and number of columns times cell resolution.
+
+```{python}
+rows = 200
+cols = 250
+resolution = 10
+tools.g_region(s=0, w=0, n=rows * resolution, e=cols * resolution, res=resolution)
+```
+
+## Creating a fractal surface as a NumPy array
+
+We will create a simple fractal surface with [r.surf.fractal](https://grass.osgeo.org/grass-stable/manuals/r.surf.fractal.html).
+
+{{< fa wand-magic-sparkles >}} The trick is to use the `output` parameter with the value `np.array` to request a NumPy array instead of a native GRASS raster. {{< fa wand-magic-sparkles >}}
+This way GRASS provides the array as the result of the call:
+
+```{python}
+import numpy as np
+
+fractal = tools.r_surf_fractal(output=np.array, seed=6)
+```
+
+::: {.callout-note title="NumPy arrays in GRASS version < 8.5"}
+
+Directly passing NumPy arrays to GRASS tools and receiving them back is a new feature in GRASS v8.5.
+If you work with older versions of GRASS, you can use
+[grass.script.array](https://grass.osgeo.org/grass-stable/manuals/libpython/grass.script.html#script.array.array) and [grass.script.array3d](https://grass.osgeo.org/grass-stable/manuals/libpython/grass.script.html#script.array.array3d):
```

Comment:
Leave out the 3D array, one, it is mentioned at the end, and two, 3D rasters/arrays are not supported at this point (easy to add, but...).

```suggestion
[grass.script.array](https://grass.osgeo.org/grass-stable/manuals/libpython/grass.script.html#script.array.array):
```